### PR TITLE
Synthetic data used in AIAU study

### PR DESCRIPTION
**Pull request by paulmorio**
## What is the goal of this PR?

The PR adds a new dataset as used in the AIAU study, with accompanying test.

### CODE REVIEW

File: pyrelational/datasets/classification/mnist.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from torch import Tensor
+from torch.utils.data import ConcatDataset
+from torchvision import datasets, transforms
+
+from pyrelational.datasets.base import BaseDataset
+
+from .utils import create_splits
+
+
+class FashionMNIST(BaseDataset):
```
Comment (a-pouplin) [inline pyrelational/datasets/classification/mnist.py:13]: Should the filename be changed to `fashion_mnist.py`?

File: requirements/base_requirements.txt
Comment (a-pouplin) [inline requirements/base_requirements.txt:1]: `rdkit>=2023.9.6` is missing

**Review by a-pouplin — DISMISSED** [review]
Nice implementation, all the tests and run files are working on my side!

File: pyrelational/datasets/classification/mnist.py (hunk 1 above)
Comment (paulmorio) [reply pyrelational/datasets/classification/mnist.py:13]: Yup that may be changed in the PR that this is currently adding to.

**Comment by paulmorio** [comment]
Pinning rdkit to 2023.9.5 passes the mypy check caused by an issue with stub generation in 2023.9.6 of rdkit (https://github.com/OpenFreeEnergy/openfe/pull/733).

File: benchmarking/PureSynthetic2D/data_manager.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class Synthetic2DDataManager(DataManager):
+    def __init__(
+        self,
+        dataset: torch.utils.data.Dataset[Tuple[torch.Tensor, torch.Tensor]],
+        random_state: int = 42,
+        numpy_flag: bool = False,
+    ):
+        self.numpy_flag = numpy_flag
```
Comment (thomasgaudelet) [inline benchmarking/PureSynthetic2D/data_manager.py:73]: does this need to be an attribute?

File: benchmarking/PureSynthetic2D/ensemble_scikit.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        Args:
+        train_loader (torch.utils.data.DataLoader): A torch dataloader with a numpy compatible collate function
+            you can also just adapt this to something else.
+        """
+        train_x, train_y = next(iter(train_loader))  # assumes dataloader returns full set of available observations
+        estimators: List[Any] = []
+
+        k = self.num_estimators
+        kf = KFold(n_splits=k)
+        from joblib import Parallel, delayed
```
Comment (thomasgaudelet) [inline benchmarking/PureSynthetic2D/ensemble_scikit.py:51]: can this be moved to top of file instead? could also used this in inference maybe?

File: pyrelational/informativeness/regression.py [hunk 4]
```diff
@@ -172,7 +172,7 @@ def _check_regression_informativeness_input(
         assert 1 <= std.ndim <= 2, "std input should be a 1D or 2D tensor"
 
 
-def _compute_mean(x: Optional[Union[Distribution, Tensor]], axis: int = 0) -> Tensor:
+def _compute_mean(x: Optional[Union[Distribution, Tensor]], axis: int = 0) -> Union[Tensor, Any]:
```
Comment (thomasgaudelet) [inline pyrelational/informativeness/regression.py:175]: hmmm, I don't think we should have Any output here, it should always be a Tensor (or nothing when raising)

File: benchmarking/PureSynthetic2D/data_manager.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+from pyrelational.data_managers import DataManager
+
+
+def numpy_collate(batch: List[Tuple[torch.Tensor, torch.Tensor]]) -> Any:
+    """Collate function for a Pytorch to Numpy DataLoader"""
+    x_list, y_list = zip(*batch)
+    return [np.stack(x_list), np.stack(y_list)]
+
+
+class Synthetic2D(Dataset[Tuple[torch.Tensor, torch.Tensor]]):
```
Comment (thomasgaudelet) [inline benchmarking/PureSynthetic2D/data_manager.py:26]: is this not a dataset in the suite? and should it be if not?

File: pyrelational/datasets/regression/aiau_synth.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import numpy as np
+import torch
+
+from pyrelational.datasets.base import BaseDataset
+from .utils import create_splits
+
+
+class AIAUSynth(BaseDataset):
+    """
+    2D example dataset used in Scherer et al. 2024 for active learning
```
Comment (thomasgaudelet) [inline pyrelational/datasets/regression/aiau_synth.py:10]: is there a full reference to paste here?

**Pull request closed without merging**